public/app/features/scenes/components/layout/SceneGridLayout.tsx

```diff
@@ -0,0 +1,438 @@
+import { css, cx } from '@emotion/css';
+import React, { useMemo } from 'react';
+import ReactGridLayout from 'react-grid-layout';
+import AutoSizer from 'react-virtualized-auto-sizer';
+
+import { GrafanaTheme2 } from '@grafana/data';
+import { Icon, useStyles2, useTheme2 } from '@grafana/ui';
+import { DEFAULT_ROW_HEIGHT, GRID_CELL_HEIGHT, GRID_CELL_VMARGIN, GRID_COLUMN_COUNT } from 'app/core/constants';
+
+import { SceneObjectBase } from '../../core/SceneObjectBase';
+import { SceneComponentProps, SceneLayoutState } from '../../core/types';
+
+interface SceneGridLayoutState extends SceneLayoutState {
+  children: Array<SceneGridCell | SceneGridRow>;
+}
+
+type GridCellLayout = {
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+};
+
+export class SceneGridLayout extends SceneObjectBase<SceneGridLayoutState> {
+  static Component = SceneGridLayoutRenderer;
+
+  updateLayout() {
+    this.setState({
+      children: [...this.state.children],
+    });
+  }
+
+  onResizeStop: ReactGridLayout.ItemCallback = (_, o, n) => {
+    const child = this.state.children.find((c) => c.state.key === n.i);
+    if (!child) {
+      return;
+    }
+    child.setState({
+      size: {
+        ...child.state.size,
+        width: n.w,
+        height: n.h,
+      },
+    });
+  };
+
+  // When layout changes, figure out if it's a nested grid layout and update enclosing cell's size if needed
+  onLayoutChange = (layout: ReactGridLayout.Layout[]) => {
+    let enclosingCell;
+    let parent = this.parent;
+    while (parent) {
+      if (parent instanceof SceneGridCell || parent instanceof SceneGridRow) {
+        enclosingCell = parent;
+        break;
+      } else {
+        parent = parent.parent;
+      }
+    }
+
+    // When not nested, we don't need to update anything
+    if (!enclosingCell) {
+      return;
+    }
```

Comment:
`onLayoutChange` is used for adjusting the sizes only for nested grids. Any other updates, i.e. children has updated x/y - is handled in respective methods (onResize... onDragStop). The `updateLayout` is then "forcing" RGL update by setting new children in the state, and hence invalidating RGL's layout.

Reply:
I don't think this is correct, it is called on mount and can change position of elements to fix overlapping elements

Reply:
> I don't think this is correct, it is called on mount and can change position of elements to fix overlapping elements

Not sure I understand. I mean, when i.e. panel is moved RGL will of course handle the layout changes - internally. But those are not "written-back" to the scene objects' state. IMO this can be solved in two ways:
1. `layout` is a state property that's basically updated whenever the children's state change. Then the `layout` needs to be provided upfront, rather than calculated based on children positions. Technically it's an optional property in the GridLayout state. 
2. As proposed in this PR `updateLayout` will update the `layout` (memoized in the renderer) when children's state is changed. 

Solution 1. is potentially tempting, but it has consequences of the API being polluted with an optional property, that only has _internal_ usage. If we had a way of specifying public state API (basically initialized via SceneObject's constructor as today) and internal SceneObject's state then it's a way to go. But I haven't yet figured out a nice typescript way to define the API this way.